### PR TITLE
reuse allocated buffer

### PR DESCRIPTION
**Pull request by tylitianrui**
**What this PR does**:
reuse  allocated  buffer  when  buffer size < 128 byte.  it  not only  reuse  allocated  buffer , but also avoid  memory leak.

### CODE REVIEW

**Review by AlexStocks — APPROVED** [review]

**Comment by wongoo** [comment]
@tylitianrui  have u meet a memory leak issue? how it happens?

**Comment by wongoo** [comment]
@tylitianrui  In concurrent scenario,  an encoder encodes an object and gets a byte buffer, but async to process it. If u reuse the buffer, it may cause the data buffer being changed by other encoding process. 

It may be better to create a new API, like `Encoder.ReuseBufferClean` , to tell user the clean will reuse the buffer.

**Comment by tylitianrui** [comment]
> @tylitianrui In concurrent scenario, an encoder encodes an object and gets a byte buffer, but async to process it. If u reuse the buffer, it may cause the data buffer being changed by other encoding process.
> 
> It may be better to create a new API, like `Encoder.ReuseBufferClean` , to tell user the clean will reuse the buffer.

YES, There is a risk of the buffer changed by other one

File: encode.go [hunk 1]
```diff
@@ -56,6 +54,22 @@ func (e *Encoder) Clean() {
 	e.refMap = make(map[unsafe.Pointer]_refElem, 7)
 }
 
+// ReuseBufferClean reuse the Encoder for a new object encoding.
+// it reuse allocated buffer and reduce memory-allocation.
+func (e *Encoder) ReuseBufferClean() {
+	var buffer []byte
+	if len(e.buffer) <= 128 {
```
Comment (wongoo) [inline encode.go:61]: does it need to check the buffer length?

**Review by wongoo — APPROVED** [review]

**Comment by wongoo** [comment]
128 may be too small for most situation，i suggest to change it to 512，or allow user to chang it by ‘hessian.SetReuseBufferSize(int)’

File: encode.go (hunk 1 above)
Comment (tylitianrui) [reply encode.go]: i  am  sorry,  i  made  a  typo.  `cap` , not  `len`.

if  not  check the  buffer capacity.  there is  a  risk of  memory leak.  for  example,  an encoder encodes an huge  object （eg: 100K ）, the allocated  room of  buffer is  also  so big.  generally, most  of   object   is  not  such huge.    it  causes most  of  buffer  room  is  not used and not    GC .  it is  

> 128 may be too small for most situation，i suggest to change it to 512，or allow user to chang it by ‘hessian.SetReuseBufferSize(int)’

good idea

**Review by AlexStocks — APPROVED** [review]

**Pull request merged**